### ISSUE TITLE
[Question] Sampling using Bernoulli Distribution

### ISSUE THREAD

**Issue by milost** [cornellius-gp/gpytorch#846]
Hi,

I'm currently working with the DKL example (https://gpytorch.readthedocs.io/en/latest/examples/08_Deep_Kernel_Learning/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.html) from the tutorials and am uncertain about the test method. To begin with, since my problem is a binary classification, I adapted the example by replacing the `SoftmaxLikelihood(num_features=model.num_dim, num_classes=num_classes)` with a simple `Bernoulli` likelihood. The training runs without any problems, I'm only unsure about what happens during the testing. I am talking about the following code snippet:

```python

def test():
    model.eval()
    likelihood.eval()

    correct = 0
    with torch.no_grad(), gpytorch.settings.num_likelihood_samples(16):
        for data, target in test_loader:
            data, target = data.cuda(), target.cuda()
            output = likelihood(model(data))  # This gives us 16 samples from the predictive distribution
            pred = output.probs.mean(0).argmax(-1)  # Taking the mean over all of the sample we've drawn
            correct += pred.eq(target.view_as(pred)).cpu().sum()
    print('Test set: Accuracy: {}/{} ({}%)'.format(
        correct, len(test_loader.dataset), 100. * correct / float(len(test_loader.dataset))
    ))
```
After the model has been switched into eval mode, `gpytorch.settings.num_likelihood_samples(16)` is used to determine how many samples we want to draw from the likelihood (here 16). The prediction is then made using `output = likelihood(model(data)))`. In the example it now says  `# This gives us 16 samples from the predictive distribution`. For this reason, the mean of the 16 samples is calculated in the next line of code. So far so good … If I do the whole thing in my case, the behavior is different though. Here the call `likelihood(model(data))` returns an object of type `Bernoulli(probs: torch.Size[64])` where `64` is the batch size. If you then execute `pred = output.probs.mean(0).argmax(-1)` in the next line, an empty tensor `tensor(0)` will be returned.  Am I doing something wrong here or does the `num_likelihood_samples(16)` setting not work when using a Bernoulli distribution?  Or is the result of the call `output = likelihood(model(data)))` already based on the 16 samples?

It would be nice if someone could clarify what’s going on … maybe point me into the right direction.

Thanks

**Comment by jacobrgardner** [comment]
When doing binary classification, you aren't going to have an additional batch dimension over the number of classes. What you are effectively doing is taking the probability prediction for each point in the minibatch, averaging those probabilities to get some constant `p`, and then taking the argmax of a singleton tensor which of course returns `0`.

Your predictions from a Bernoulli model are just `(output.probs > 0.5).int()`.

**Comment by milost** [comment]
Hey Jake, thanks for the clarification! Now I get it.
